File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ItemPriceInputString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ItemPriceInputString {
    private static final String DNCString = "?DNC?";

    public String totalItems;

    public ItemPriceInputString(String totalItems) {
        this.totalItems = totalItems;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static ItemPriceInputString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "ItemPriceInput");
        return new ItemPriceInputString(parts.get(0));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ItemPriceInputString)) return false;
        ItemPriceInputString that = (ItemPriceInputString) o;
        return (DNCString.equals(totalItems) || DNCString.equals(that.totalItems) || Objects.equals(totalItems, that.totalItems));
    }

    @Override
    public int hashCode() {
        return Objects.hash(totalItems);
    }

    @Override
    public String toString() {
        return Tokens.token(totalItems);
    }
}

File: JavaScript/ExampleTests/common/ItemPriceInputString.js
import * as tokens from "./tokens.js";

export class ItemPriceInputString {
  static DNC_STRING = "?DNC?";

  constructor(totalItems = "") {
    this.totalItems = totalItems;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ItemPriceInputString(
      v[0] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 1, "ItemPriceInput");
    return new ItemPriceInputString(
      parts[0]
    );
  }

  toString() {
    return tokens.token(this.totalItems);
  }

  equals(other) {
    if (!(other instanceof ItemPriceInputString)) return false;
    return (this.totalItems === ItemPriceInputString.DNC_STRING || other.totalItems === ItemPriceInputString.DNC_STRING || this.totalItems === other.totalItems);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
